### PR TITLE
[WIP] Firefox compatibility

### LINKED ISSUE #43: Port to Firefox
Current nighty versions of Firefox 57 have already disabled legacy (XUL) add-ons and once the version is released in the stable channel (November 2017), the current extension for Firefox will stop working, as only WebExtensions will be supported. The best way to address it would be to port this to Firefox, hopefully only repacking will be needed. I will take a look on it in the coming days.

### PR DESCRIPTION
**Pull request by MikkCZ**
This should mostly fix #43. But there are still few things to do.
- [ ] add actual link to AMO reviews
- [x] set up build for Chrome and Firefox (two different `manifest.json` files are needed)
- [x] check if it's possible to open 'about:preferences#sync' instead of support.mozilla.org article ([it's not](https://discourse.mozilla.org/t/add-a-link-to-about-preferences-sync/18086))
- [x] fix the context menu
- [x] fix changelog link
- [x] reviews link should display AMO in Firefox
- [x] sync service name is Firefox Sync
- [x] sync does not work in Firefox without ID in manifest.json

### CODE REVIEW

File: source/js/options.js [hunk 1]
```diff
@@ -129,20 +129,27 @@ $(document).ready(function() {
...
-    $("#changelogText").text(text);
-  });
+  var changeLogRequest = new XMLHttpRequest();
+  changeLogRequest.onreadystatechange = function() {
+    if (changeLogRequest.responseText) {
+      $("#changelogText").text(changeLogRequest.responseText);
+    }
+  }
+  changeLogRequest.open('GET', 'ChangeLog', true);
+  changeLogRequest.send();
```
Comment (kav2k) [inline source/js/options.js]: Why is this change needed?

Also, if we're to replace `$.ajax` anywhere, let's be modern and use [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or at the very least `load` event instead of `onreadystatechange`.

File: source/js/options.js [hunk 2]
```diff
@@ -101,6 +101,18 @@ function openSyncSettings() {
...
 
+function displaySyncSettingsLink() {
+  if (typeof browser !== 'undefined' || browser !== null) {
+    // We are on Firefox
+    $('#open_sync_settings').text('Firefox Sync');
+    $('#open_sync_settings').attr('href', 'https://support.mozilla.org/kb/how-do-i-choose-what-types-information-sync-firefox');
+  } else {
+    // We are in Chrome
+    $('#open_sync_settings').text('Chrome Sync');
+    $('#open_sync_settings').click(openSyncSettings);
```
Comment (kav2k) [inline source/js/options.js]: Refactor name as `openChromeSyncSettings` for clarity.

File: source/manifest.json [hunk 3]
```diff
@@ -1,4 +1,10 @@
 {
+  "applications": {
+    "gecko": {
+      "id": "theoldreader@knyar",
+      "strict_min_version": "53.0"
+    }
+  },
```
Comment (kav2k) [inline source/manifest.json]: If I remember correctly, this will fail to validate in Chrome; also, might not be needed in FF (for publishing).
But my knowledge of FF is shaky here.
Split manifest is one of the reasons I am wary of a 100% unified codebase.

File: source/js/options.js [hunk 4]
```diff
@@ -101,23 +101,55 @@ function openSyncSettings() {
...
+  }
+}
+
+function showReviewsLink() {
+  if (typeof browser !== 'undefined') {
+    // We are on Firefox
+    $('#reviewsLink').text('Write a review on Addons.Mozilla.org');
```
Comment (kav2k) [inline source/js/options.js]: "Mozilla Add-Ons", [as per docs](https://wiki.mozilla.org/AMO).

File: source/js/options.js [hunk 5]
```diff
@@ -101,23 +101,55 @@ function openSyncSettings() {
   chrome.tabs.create({url: "chrome://settings/syncSetup"});
 }
 
+function displaySyncSettingsLink() {
+  if (typeof browser !== 'undefined') {
```
Comment (kav2k) [inline source/js/options.js]: Browser detection needs to be factored out in a separate function.

File: source/js/options.js (hunk 1 above)
Comment (MikkCZ) [inline source/js/options.js]: I haven't found any details, but `$.ajax` fails. On stackoverflow I have found only something about CSP for injected scripts, which is not the case here. I expect it does not use only `XMLHttpRequest` on the background and gets blocked. Using `fetch()` sounds reasonable.

File: source/manifest.json (hunk 3 above)
Comment (MikkCZ) [inline source/manifest.json]: It's necessary for [`storage.sync`](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage/sync). If it's the only difference, that could be easily solved with a simple build script and separate manifests. Or we can drop sync support for Firefox, what sound better for you.

**Comment by MikkCZ** [comment]
The context menu was actually not broken. I thought it should be displayed for the toolbar button. But when looking at the code, I see it should be for the page context only.

File: source/manifest.json (hunk 3 above)
Comment (kav2k) [inline source/manifest.json]: I think a build script would be the best solution.

Comment (MikkCZ) [inline source/manifest.json]: Do you already have a build script for Chrome? Can you add it here? To be honest, I have no experience with Chrome add-ons packaging.

Comment (kav2k) [inline source/manifest.json]: We don't need to do any complex packaging; generating `.crx` is not needed, just copying stuff over to `.gitignore`d folder like `build` should be enough.

Though even better would be to also produce a version-stamped `zip` for upload.

The tricky question is making this build script cross-platform (yes, I do a lot on Windows, sadly). Perhaps using some build system like Grunt would make the _pain_ equal cross-platform.. 😛

Comment (MikkCZ) [inline source/manifest.json]: This is where we hit my knowledge. Never worked with npm or grunt. But https://www.npmjs.com/package/grunt-webext-builder and https://www.npmjs.com/package/grunt-crx look promising. @kav2k if you have experience with grunt, can you try those?

Comment (kav2k) [inline source/manifest.json]: I think this requires more research and can be done outside this PR. For now, just a separate manifest file + `build` folder + `build/*` in `.gitignore` is sufficient.

**Comment by MikkCZ** [comment]
I need to look at Window batch scripts later, once I get to a machine, where I can test it.

**Comment by kav2k** [comment]
I don't think a separate build system for Windows is needed right now. After all, WSL is a thing on Win10, so using the Makefile isn't out of the question. I'm ready to merge as-is.

**Comment by kav2k** [comment]
Remaining problems after some testing:

1. Build script doesn't remove the other manifest from the package.
2. Bookmark context menu entries don't work. I'll investigate.
3. Firefox sync hint should open in a new tab instead of navigating the options frame.
4. <kbd>B</kbd> shortcut in TOR interface ("open in background") doesn't work.

**Comment by kav2k** [comment]
5 can be fixed by replacing `selected` with `active` in `fuctions.js:56`.

**Comment by kav2k** [comment]
Hm, and now I can't reproduce 4..

**Comment by kav2k** [comment]
The bookmark code seems to be a mess. I'd be ok if you just conditionally added the context menus for now.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+CP_R=/bin/cp -r
+ZIPPROG=/usr/bin/zip
+RM_R=/bin/rm -r
+
+zip_all :: zip_chrome zip_firefox
+
+zip_chrome:
+	$(MKDIR) -p $(TARGET)/chrome && \
+	$(CP_R) $(SRC)/* $(TARGET)/chrome && cd $(TARGET)/chrome && \
+	$(MV) manifest-chrome.json manifest.json && \
```
Comment (kav2k) [inline Makefile]: Should also remove the other manifest.

File: source/js/options.js [hunk 7]
```diff
@@ -96,37 +104,64 @@ function show_message(message) {
...
+function showReviewsLink() {
+  switch (getBrowserName()) {
+    case 'Mozilla':
+      $('#reviewsLink').text('Write a review on Mozilla Add-Ons');
+      $('#reviewsLink').attr('href', 'https://addons.mozilla.org/addon/this-add-on-id/reviews');
+      break;
+    case 'Chrome':
+    default:
+      $('#reviewsLink').text('Rate in Chrome Web Store');
+      $('#reviewsLink').attr('href', 'https://chrome.google.com/webstore/detail/the-old-reader-notifier/flnadglecinohkbmdpeooblldjpaimpo/reviews');
```
Comment (kav2k) [inline source/js/options.js]: If we're using a plain link here, add `target="_blank"` (or use `chrome.tabs`).

**Comment by MikkCZ** [comment]
I need to apologize, I will be away for about a week from now. Maybe I will have some time tomorrow, but if not, feel free to pick this up and continue, else I will continue once I am back again. ;)

**Comment by kav2k** [comment]
I'm going to merge this into a separate branch here and continue working on it. I'll split off issues - feel free to take some on if you have time.
